broadinstitute/single_cell_portal_core
# Find any gene by name in differential expression table (SCP-4814)

This adds a search box to the DE table, which lets users find any gene in the table by its name.

Beyond seeing the 15 most differentially expressed genes, users can now quickly and easily learn if their gene of interest is significantly differentially expressed, and if so by how much.  Or they can see if genes with similar names to those listed by default -- i.e., often their paralogs -- are also very differentially expressed.

### Video <span id="video"></span>
Here's how it looks:

https://user-images.githubusercontent.com/1334561/216365576-b0ddce42-951e-47ba-98cb-b3f7ee2be99e.mov

### Analytics <span id="analytics"></span>
Instrumentation on this new search feature aligns with our other search usage metrics.  It lets us answer many questions, e.g.:
* How much usage does this feature get?
* What gene was searched in the DE table?
* Does usage differ by species?
* How many times did the user do this search on this page view?

Notably, while a search is done after each key press, a search is only logged as of 1 second after the last key press.  So typing e.g. "PIK" in relatively quick succession would log only 1 search, not 3, even though the UI technically updated instantly with 3 sets of search results.  We're really only interested in analyzing interactions where the user could reasonably be inferred to visually process the results -- not ultra-short inadvertent or in-transit interactions where the user does not look at what the UI responds with.  This improves the signal-to-noise ratio for our analytics.  The "delayed logging" technique introduced here might help denoise similar analytics, like scatter plot legend hover logging.

Check out the [Mixpanel report preview](https://mixpanel.com/s/2NikvP).  The same query will populate the [production Mixpanel report](https://mixpanel.com/s/Sq4YS) after release.  You can also find it from the [differential expression analytics board](https://mixpanel.com/s/2ZxI7z).

### Test <span id="test"></span>
An automated test confirms behavior.  Optionally, if you want to manually verify:
1.  **If local**, after pulling, since we're only testing frontend changes, try booting with the frontend service worker cache enabled for faster page loads: `bin/docker-compose-setup.sh -c`.  Then go to a study with DE results.  **If on staging**, go to ["Human milk - differential expression"](https://singlecell-staging.broadinstitute.org/single_cell/study/SCP303/human-milk-differential-expression#study-visualize) (SCP303).
2. Click the blue "Differential expression" button
3. Select an annotation group from the drop-down
4. Type in the "Find gene" search box
5. Confirm results have a gene name that includes your queried text
6. Click the "x" to clear results
7. Open DevTools, go to Network panel
8. Add "method:POST bard" to the filter at upper left
9. In "Find gene", type 3 characters quickly, like a user typically would
10. In DevTools Network panel, confirm only 1 new post to Bard for "search" was sent

If you see an error in DevTools Console like the following, please disregard:
```
Warning: A component is changing an uncontrolled input of type radio to be controlled. Input elements should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled input element for the lifetime of the component. More info: https://fb.me/react-controlled-components
    in input (at DifferentialExpressionPanel.jsx:107)
    ...
```

This seems to be a pre-existing issue, and has likely existed since the DE table was created.  I spent much of a day trying to fix it without success.  It's reproducible for Emily and me on local SCP on `development`, but not staging with the same code.  It also has no user impact, so it seems reasonable to not invest more time here to resolve it.

This satisfies SCP-4814.
